cpp/src/arrow/flight/client_header_internal.h

```diff
@@ -22,6 +22,8 @@
 
 #include "arrow/flight/client_middleware.h"
 #include "arrow/result.h"
+#include "arrow/util/optional.h"
```

Comment:
nit: I don't think value_parsing is used in the header, and we're missing `#include <mutex>`, and `#include <functional>` for `binary_function`, `#include <string>`, and `#include <chrono>`

Reply:
Oh and `#include <map>` but perhaps unordered_map is better since you don't care about ordering?